src/app_selection.rs

```diff
@@ -0,0 +1,353 @@
+use holochain_client::AppInfo;
+use holochain_conductor_api::CellInfo;
+use holochain_types::dna::DnaHash;
+use std::ops::Deref;
+use std::sync::Arc;
+use thiserror::Error;
+
+use crate::{config::AllowedAppIds, AdminCall};
+
+#[derive(Debug, PartialEq, Error)]
+pub enum AppSelectionError {
+    #[error("App is not installed on the conductor")]
+    NotInstalled,
+
+    #[error("App is not in the list of allowed apps")]
+    NotAllowed,
+
+    #[error("Multiple matching apps were found, could not determine which to call")]
+    MultipleMatching,
+}
+
+pub type AppInfoCache = Arc<tokio::sync::RwLock<Vec<AppInfo>>>;
+
+/// Return the [`AppInfo`] of the matching valid app if unique.
+///
+/// The returned app must meet the following criteria:
+/// - It contains a cell with the given `dna_hash`.
+/// - It can be identified by the given `coordinator_identifier`.
+/// - It is in the list of `installed_apps` configured for the gateway.
+///
+/// If a unique app matching the criteria cannot be found, then an error is returned.
+///
+/// # Side effects
+/// If a matching app is not found in the provided list of installed apps then a request to the
+/// admin websocket will be made and the list will be updated with the results of that request.
+pub async fn try_get_valid_app(
+    dna_hash: DnaHash,
+    coordinator_identifier: String,
+    installed_apps: AppInfoCache,
+    allowed_apps: &AllowedAppIds,
+    admin_call: impl Deref<Target = impl AdminCall + ?Sized>,
+) -> Result<AppInfo, AppSelectionError> {
+    let app_info = {
+        let installed_apps = installed_apps.read().await;
+        choose_unique_app(&dna_hash, &coordinator_identifier, &installed_apps)
+            .ok()
+            .cloned()
+    };
+
+    let app_info = match app_info {
+        Some(app_info) => app_info,
+        None => {
+            let mut installed_apps = installed_apps.write().await;
```

Comment:
Should we move this write lock down to where it is actually used, rather than before waiting for a admin ws call?

Reply:
In this case yes, it may as well. We aren't actually preventing repeated lookups here so we may as well be light on the lock

I'll do this in a follow up

Reply:
See https://github.com/holochain/hc-http-gw/pull/31